src/modules/hip/kernel/non_silent_region_detection.hpp

```diff
@@ -0,0 +1,401 @@
+        __syncthreads();
+    }
+
+    if (hipThreadIdx_x == 0)
+    {
+        int dstIdx = id_z * hipGridDim_x + hipBlockIdx_x;
+        maxArr[dstIdx] = max_smem[0];
+    }
+}
+
+__global__ void cutoffmag_hip_tensor(float *srcPtr,
+                                     int maxLength,
+                                     float *cutOffMagPtr,
+                                     float cutOff,
+                                     float referencePower,
+                                     bool referenceMax)
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+
+    // if referenceMax is set to true, perform final max reduction on srcPtr and compute cutOffMag
+    if(referenceMax)
+    {
+        uint srcIdx = id_z * maxLength;
+        __shared__ float max_smem[256];
+        max_smem[hipThreadIdx_x] = srcPtr[srcIdx];
+
+        if (id_x >= maxLength)
+            return;
+
+        srcIdx += id_x;
+        float maxVal = srcPtr[srcIdx];
+        while (id_x < maxLength)
+        {
+            maxVal = fmaxf(maxVal, srcPtr[srcIdx]);
+            id_x += hipBlockDim_x;
+            srcIdx += hipBlockDim_x;
+        }
+        max_smem[hipThreadIdx_x] = maxVal;
+        __syncthreads();
+
+        // do reduction on min_smem and max_smem
+        for (int threadMax = 128; threadMax >= 1; threadMax /= 2)
+        {
+            if (hipThreadIdx_x < threadMax)
+                max_smem[hipThreadIdx_x] = max(max_smem[hipThreadIdx_x], max_smem[hipThreadIdx_x + threadMax]);
+            __syncthreads();
+        }
+
+        // cutOffMag = max(srcPtr) * cutoff;
+        if (hipThreadIdx_x == 0)
+            cutOffMagPtr[id_z] = max_smem[0] * cutOff;
+    }
+    else
+    {
+        if (hipThreadIdx_x == 0)
+            cutOffMagPtr[id_z] = referencePower * cutOff;
+    }
+}
+
+// -------------------- Set 3 -  kernels for finding begin and length of NSR in inputs --------------------
+
+__global__ void find_region_hip_tensor(float *srcPtr,
+                                       uint nStride,
+                                       int *beginTensor,
+                                       int *lengthTensor,
+                                       float *cutOffMagPtr,
+                                       int *srcLengthTensor,
+                                       float windowLength)
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+    uint srcLength = srcLengthTensor[id_z];
+    float cutOffMag = cutOffMagPtr[id_z];
+
+    __shared__ int beginResult;
+    __shared__ int endResult;
+    beginResult = srcLength;
+    endResult = 0;
+    __syncthreads();
+
+    int beginIdx = srcLength;
+    int endIdx = 0;
+    for (int i = id_x; i < srcLength; i += hipBlockDim_x)
+    {
+        uint srcIdx = id_z * nStride + i;
```

Comment:
id_z * nStride is loop independent

Reply:
Loop can directly run from (id_z * nStride) + id_x till (id_z * nStride) + srcLength so you can remove L221 inside the loop